### PR TITLE
Feat/dwa

### LINKED ISSUE #66: Convert Vue PWA (Progressive Web App) to a DWA (Decentralized Web App)
# Convert Vue PWA to a Decentralized Web App (DWA) 

## 🚀 Goal
Begin the process of transforming our Vue.js Progressive Web App (PWA) into a Decentralized Web App (DWA) by setting up the necessary Web5 infrastructure.

## 🤔 Background
A Decentralized Web App (DWA) is a Progressive Web App (PWA) that incorporates Web5 functionalities, providing offline capabilities and offline synchronization of Decentralized Web Nodes (DWNs). This task will initiate the conversion of our existing Vue PWA to a DWA by setting up the basic Web5 components.

> This is part of our larger project to create a Vue.js DWA starter. See our [main issue here](https://github.com/TBD54566975/tbd-examples/issues/63) for the full context and list of all related tasks.

> Important: For reference, please see the [DWA React Vite starter app](https://github.com/TBD54566975/tbd-examples/tree/main/javascript/dwa-starter-react-vite). While the implementation details will differ for Vue, this example provides a good overview of the structure and functionalities of a DWA.

## 🔑 Tasks and Acceptance Criteria
- [ ] Install the `@web5/api` package https://www.npmjs.com/package/@web5/api
- [ ] Import and implement `activatePolyfills` from `@web5/api` in the main application file
- [ ] Update the service worker to include Web5 polyfills:
  - [ ] Import `activatePolyfills` in the service worker
  - [ ] Implement `activatePolyfills` with appropriate cache TTL settings
- [ ] Verify that the app still functions as a PWA after these changes

## 🌟 Resources
- [Web5 DWA documentation](https://developer.tbd.website/docs/web5/decentralized-web-nodes/access-data-with-drls/)
- [DWA Starter App Example](https://github.com/TBD54566975/tbd-examples/tree/main/javascript/dwa-starter-react-vite)
- [Vue.js Documentation](https://vuejs.org/guide/introduction.html)
- [Progressive Web Apps (PWAs) | MDN](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps)

### Importing and Activating Polyfills
```javascript
import { activatePolyfills } from '@web5/api';

// Activate Web5 Service Worker Polyfills
activatePolyfills({
  onCacheCheck() {
    return {
      ttl: 30000,
    };
  },
});
```

### Service Worker Configuration (example)
```javascript
/// <reference lib="webworker" />
import {
  cleanupOutdatedCaches,
  createHandlerBoundToURL,
  precacheAndRoute,
} from "workbox-precaching";
import { NavigationRoute, registerRoute } from "workbox-routing";
import { activatePolyfills } from "@web5/api";

declare let self: ServiceWorkerGlobalScope;

self.addEventListener("message", (event) => {
  if (event.data && event.data.type === "SKIP_WAITING") self.skipWaiting();
});

precacheAndRoute(self.__WB_MANIFEST);
cleanupOutdatedCaches();

let allowlist;
if (import.meta.env.DEV) allowlist = [/^\/$/];

registerRoute(
  new NavigationRoute(createHandlerBoundToURL("index.html"), { allowlist })
);

activatePolyfills({
  onCacheCheck() {
    return {
      ttl: 30000,
    };
  },
});
```

## Getting Started
1. Comment ".take" on this issue to get assigned
2. Fork the repository and create a new branch for this task
3. Follow the tasks outlined above
4. Submit a pull request with your changes
5. Respond to any feedback during the review process

## Questions?
If you have any questions or need clarification, please comment on this issue or join our [Discord](https://github.com/TBD54566975/tbd-examples/issues/64#) community.

Happy coding! 🎉

### PR DESCRIPTION
**Pull request by Johnnyevans32**
## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ New Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 👷 Example Application
- [ ] 🧑‍💻 Code Snippet
- [ ] 🎨 Design
- [ ] 📖 Content
- [ ] 🧪 Tests
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description

This PR adds Web5 connection converting it to a DWA. 

## Related Tickets & Documents

Resolves #66 

## Mobile & Desktop Screenshots/Recordings

https://github.com/user-attachments/assets/0d37001f-d1a1-47c3-bf6c-1efd7f3ec68c

## Added code snippets?
- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 readme
- [ ] 📜 contributing.md
- [ ] 📓 general documentation
- [x] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
thanos-thano

### CODE REVIEW

**Comment by acekyd** [comment]
Checking this out. Thanks for the contribution @Johnnyevans32

File: javascript/dwa-starter-vue/src/composables/web5.ts
Comment (leordev) [inline javascript/dwa-starter-vue/src/composables/web5.ts:16]: since they are both framework agnostic, can you please maintain the original files below and connect with this composable?
- https://github.com/TBD54566975/tbd-examples/blob/main/javascript/dwa-starter-react-vite/src/lib/todo-dwn-repository.ts
- https://github.com/TBD54566975/tbd-examples/blob/main/javascript/dwa-starter-react-vite/src/web5/protocols.ts

this will make things simpler to maintain across all the dwa-starter examples

File: javascript/dwa-starter-vue/e2e/vue.spec.ts
Comment (leordev) [inline javascript/dwa-starter-vue/e2e/vue.spec.ts:17]: do you need help with this test or are you planning to add it back in one of the other projects tasks?

Comment (Johnnyevans32) [reply javascript/dwa-starter-vue/e2e/vue.spec.ts:17]: yeah I do actually, i was going to get more info on this after finishing up the main pages components

Comment (leordev) [reply javascript/dwa-starter-vue/e2e/vue.spec.ts:17]: lets do this... hold off on this test for now, and lets merge this pr as soon as you address my points above. then after that I can work on mocking/setting up the connection in playwright while you work on your next task. this way we dont block you!

Comment (Johnnyevans32) [reply javascript/dwa-starter-vue/e2e/vue.spec.ts:17]: gotcha thanks!

File: javascript/dwa-starter-vue/src/composables/web5.ts
Comment (Johnnyevans32) [reply javascript/dwa-starter-vue/src/composables/web5.ts]: done with this @leordev, I found the react implementation of the web5 composable a bit confusing hence why I opted for a different version

Comment (leordev) [reply javascript/dwa-starter-vue/src/composables/web5.ts]: What I suggested was to copy these two files to your project and use them in the `useWeb5` or in the best way you can think of. Is it possible?

Comment (Johnnyevans32) [reply javascript/dwa-starter-vue/src/composables/web5.ts]: ah ok got that now

Comment (Johnnyevans32) [reply javascript/dwa-starter-vue/src/composables/web5.ts]: should be ok now i hope @leordev

**Comment by Johnnyevans32** [comment]
hi @leordev, have you had the time to look at the updates I made, if there are additional changes I need to make.

**Comment by blackgirlbytes** [comment]
@Johnnyevans32 i tested and this is ready to go..the only thing that needs to be fixed before I approve and merge is the conflict on the pnpm lock file. Thank you!

**Comment by Johnnyevans32** [comment]
> @Johnnyevans32 i tested and this is ready to go..the only thing that needs to be fixed before I approve and merge is the conflict on the pnpm lock file. Thank you!

sorted now @blackgirlbytes

**Review by blackgirlbytes — APPROVED** [review]
Thanks for the contribution! This was a tough one.

**Comment by blackgirlbytes** [comment]
Hey @taniashiba ..after we remove the leaderboard action from github..we need to manually add ANOTHER 10 points for @Johnnyevans32 to the leaderboard because this PR he did two issues in one..a large and a medium. So instead of having 115..he would have 115 + 10 (from other PR) + 10 (from this PR) = 135

**Pull request merged**
